### DISCUSSION TITLE
Rewrite "select" operand

### DISCUSSION THREAD

**Discussion by alexsokolsky** [WebAssembly/binaryen#5973] [Q&A]
I found a bug in one of the webassembly implementations. As a workaround I need to replace certain "select" operands with "if" operands.
 Is there a snippet somewhere that does this (assuming the branches contain side effects) ?

**Comment by kripken** [comment]
Here is code that does the opposite (`if => select`) that might be useful: https://github.com/WebAssembly/binaryen/blob/4f1295a96d175f39257a750e06a426e5beb3933a/src/passes/RemoveUnusedBrs.cpp#L1247

And here is code that turns an `or` into an `if`: https://github.com/WebAssembly/binaryen/blob/4f1295a96d175f39257a750e06a426e5beb3933a/src/passes/OptimizeInstructions.cpp#L3082

**Reply by bulocode** [reply]
@kripken, both cases - have side effects check, smth like
```
EffectAnalyzer ifFalse(passOptions, *getModule(), iff->ifFalse);
if (ifFalse.hasSideEffects()) {
    return nullptr;
}
```
Are there any workarounds, to replace **select & side effects** with **if** ?
Maybe it is possible - to skip some side effects - implicitTrap or readsMemory...

**Reply by kripken** [reply]
That might be possible, and we already ignore traps if trapsNeverHappen mode is enabled, for example. There might be more cases we could do this, but I can't think of any atm.
